File: sliding-window/permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1)>len(s2): return False

        s1count = [0]*26
        s2count = [0]*26

        for i in range(len(s1)):
            s1count[ord(s1[i]) - ord('a')]+=1
            s2count[ord(s2[i]) - ord('a')]+=1
        
        match = sum(map(lambda i: s1count[i] == s2count[i], range(26)))

        for i in range(len(s2)-len(s1)):
            if match == 26:
                return True
            
            #remove the left side
            s2count[ord(s2[i])-ord('a')] -=1
            if s2count[ord(s2[i]) - ord('a')] == s1count[ord(s2[i]) - ord('a')]:
                match+=1
            elif s2count[ord(s2[i])-ord('a')]+1 == s1count[ord(s2[i]) - ord('a')]:
                match-=1

            #add the right side
            r = s2[i+len(s1)]
            s2count[ord(r)-ord('a')] +=1
            if s2count[ord(r) - ord('a')] == s1count[ord(r) - ord('a')]:
                match+=1
            elif s2count[ord(r)-ord('a')]-1 == s1count[ord(r) - ord('a')]:
                match-=1
        
        return match==26

File: hash-table/permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        // Sliding window
        if (s1.length() > s2.length()){
            return false; 
        }
        HashMap<Character, Integer> s1Count = new HashMap<>(); 
        HashMap<Character, Integer> s2Count = new HashMap<>(); 

        for(int i = 0; i < s1.length(); i++){
            s1Count.put(s1.charAt(i), s1Count.getOrDefault(s1.charAt(i), 0) + 1);
            s2Count.put(s2.charAt(i), s2Count.getOrDefault(s2.charAt(i), 0) + 1);
        }

        if (s2Count.equals(s1Count)){
            return true; 
        }

        // the s2Count maintains a window size of s1.length(); 
        int left = 0; 
        for(int j = s1.length(); j < s2.length(); j++){
            char curChar = s2.charAt(j); 
            s2Count.put(curChar, s2Count.getOrDefault(curChar, 0) + 1);
            
            char leftChar = s2.charAt(left); 
            s2Count.put(leftChar, s2Count.get(leftChar) - 1);
            if (s2Count.get(leftChar) == 0){
                s2Count.remove(leftChar); 
            }
            left++; 

            if (s2Count.equals(s1Count)){
                return true;
            }
        }

        return false; 
    }
}